src/core/core/address.cpp

```diff
@@ -153,8 +155,91 @@ DSN_API uint32_t dsn_ipv4_local(const char *network_interface)
     return ret;
 }
 
+rpc_address::~rpc_address() { clear(); }
+
+rpc_address::rpc_address(const rpc_address &another) { *this = another; }
+
+rpc_address &rpc_address::operator=(const rpc_address &another)
+{
+    if (this == &another) {
+        // avoid memory leak
+        return *this;
+    }
+    clear();
+    _addr = another._addr;
+    switch (another.type()) {
+    case HOST_TYPE_GROUP:
+        group_address()->add_ref();
+        break;
+    case HOST_TYPE_URI:
+        uri_address()->add_ref();
+        break;
+    default:
+        break;
+    }
+    return *this;
+}
+
+void rpc_address::assign_uri(const char *host_uri)
+{
+    clear();
+    _addr.uri.type = HOST_TYPE_URI;
+    dsn::rpc_uri_address *addr = new dsn::rpc_uri_address(host_uri);
+    // take the lifetime of rpc_uri_address, release_ref when change value or call deconstruct
+    addr->add_ref();
+    _addr.uri.uri = (uint64_t)addr;
+}
+
+void rpc_address::assign_group(const char *name)
+{
+    clear();
+    _addr.group.type = HOST_TYPE_GROUP;
+    dsn::rpc_group_address *addr = new dsn::rpc_group_address(name);
+    // take the lifetime of rpc_uri_address, release_ref when change value or call deconstruct
+    addr->add_ref();
+    _addr.group.group = (uint64_t)addr;
+}
+
+rpc_address rpc_address::clone() const
```

Comment:
I didn't see any difference between `clone` and `operator=`. Even if they are implemented differently, they are supposed to be the same. Remove `clone`?

Reply:
clone is used for uri_resolver in the old code, currently it's indeed useless. Let's remove it.